test/Test/Database/LSMTree/Internal/Snapshot/Codec/Golden.hs

```diff
@@ -29,328 +30,453 @@ import           System.FS.API.Types (FsPath, MountPoint (..), fsToFilePath,
+  Enumeration class
+-------------------------------------------------------------------------------}
+
+-- | Enumerate values of type @a@ for golden tests
+--
+-- To prevent combinatorial explosion, the enumeration should generally be
+-- /shallow/: the different constructors for type @a@ should be enumerated
+-- without recursively enumerating the constructors' fields. For example,
+-- enumerating @Maybe Int@ should give us something like:
+--
+-- > enumGolden @(Maybe Int) = [ Just 17, Nothing ]
+--
+-- This is generally a suitable approach, since the snapshot encodings do not
+-- encode types differently depending on values in the constructor fields.
+--
+-- Example (recursive) instances that prevent combinatorial explosion:
+--
+-- @
+--  instance EnumGolden a => EnumGolden (Maybe a) where
+--    enumGolden = [ Just (singGolden @a), Nothing ]
+--  instance EnumGolden Int where
+--    enumGolden = [17, -72] -- singGolden = 17
+-- @
+--
+-- If there are encoders that do require more elaborate (recursive)
+-- enumerations, then it is okay to deviate from shallow enumerations, but take
+-- care not to explode the combinatorics ;)
+class EnumGolden a where
+  {-# MINIMAL enumGolden | enumGoldenAnnotated | singGolden #-}
+
+  -- | Enumerated values. The enumeration should be /shallow/.
+  --
+  -- The default implementation is to return a singleton list containing
+  -- 'singGolden'.
+  enumGolden :: [a]
+  enumGolden = [ singGolden ]
+
+  -- | Enumerated values with an annotation for naming purposes. The enumeration
+  -- should be /shallow/, and the annotations should be unique.
+  --
+  -- The default implementation is to annotate 'enumGolden' with capital letters
+  -- starting with @\'A\'@.
+  enumGoldenAnnotated :: [(Annotation, a)]
+  enumGoldenAnnotated = zip [[c] | c <- ['A' .. 'Z']] enumGolden
+
+  -- | A singleton enumerated value. This is mainly useful for superclass
+  -- instances.
+  --
+  -- The default implementation is to take the 'head' of 'enumGoldenAnnotated'.
+  singGolden :: a
+  singGolden = snd $ head enumGoldenAnnotated
+
+type Annotation = String
+
+enumGoldenAnnotated' :: EnumGolden a => Proxy a -> [(Annotation, a)]
+enumGoldenAnnotated' _ = enumGoldenAnnotated
+
+{-------------------------------------------------------------------------------
+  Enumeration class: names and file paths
+-------------------------------------------------------------------------------}
+
+nameGolden :: Typeable a => Proxy a -> Annotation -> String
+nameGolden p ann = map spaceToUnderscore (show $ typeRep p) ++ "." ++ ann
+
+spaceToUnderscore :: Char -> Char
+spaceToUnderscore ' ' = '_'
+spaceToUnderscore c   = c
+
+{-------------------------------------------------------------------------------
+  Enumeration class: instances
+-------------------------------------------------------------------------------}
+
+instance EnumGolden SnapshotMetaData where
+  singGolden = SnapshotMetaData singGolden singGolden singGolden singGolden singGolden
```

Comment:
This pattern of applying a constructor to a string of `singGoldens` looks like it could be generically derived. Maybe I should add a TODO